### PR TITLE
Epng 3494 add async validation to form

### PR DESCRIPTION
**Pull request by asandau**
This is probably not up to your standards. But it seems to work :D
Please review and give suggestions on how to improve. Thanks

### CODE REVIEW

File: src/form/Form.jsx [hunk 1]
```diff
@@ -97,13 +106,28 @@ export default class Form extends React.Component {
...
+        errors: containsError(errors) ? errors : new Immutable.Map()
+      })
+    }
     this.setState({
-      errors: containsError(errors) ? errors : new Immutable.Map(),
       value: newValue2 || newValue1
     })
   }
 
+  handleAsyncValidate = (value, changedFiled, triedToSubmit) => {
```
Comment (ingmarh) [inline src/form/Form.jsx]: For what is this new argument, `changedFiled`, supposed to be used?

File: src/form/Form.jsx [hunk 2]
```diff
@@ -123,24 +147,34 @@ export default class Form extends React.Component {
...
+    } else {
+      return new Promise((resolve, reject) => {
+        resolve(this.props.validate(value, triedToSubmit))
+      })
     }
   }
 
   render () {
-    const {name, value, onSubmit, onChange, prepare, validate, normalize, disabled, children, ...other} = this.props // eslint-disable-line no-unused-vars
+    const {name, value, onSubmit, onChange, prepare, validate, validateAsync, validateAsyncWaitMs, normalize, disabled, children, ...other} = this.props // eslint-disable-line no-unused-vars
```
Comment (ingmarh) [inline src/form/Form.jsx]: Instead of the comment at the end of this line, please use `// eslint-disable-next-line no-unused-vars` one line above.

**Review by ingmarh — CHANGES_REQUESTED** [review]
Nice addition!

I was thinking about if we can have the "Promise linearization" without (the copied into) `createPromiseLinearizer` function.

I was also thinking about the API additions: Instead of a new `validateAsync`, I'd prefer to extend the existing `validate` prop handling with async support and have the new `validateAsyncWaitMs` prop renamed to `validateWaitMs`.

Something like this:

```jsx
function createLinearizedDebouncedFunc (func, waitMs = 0) {
  let timer = null

  return (...args) => {
    return new Promise(resolve => {
      if (timer) window.clearTimeout(timer)
      timer = window.setTimeout(() => {
        timer = null
        resolve(func(...args))
      }, waitMs)
    })
  }
}

export default class Form extends React.Component {
  static propTypes = {
    // ...    
    validateWaitMs: PropTypes.number,
  }

  static defaultProps = {
    // ...
    validateWaitMs: 0,
  }

  constructor (props) {
    super(props)

    // supports `props.validate(value, forSubmit): object | Promise<object>`
    this.validate = createLinearizedDebouncedFunc(props.validate, props.validateWaitMs)
  }

  setValue = (name, value) => {
    // ...
    const newValue = newValue2 || newValue1

    this.setState({ value: newValue })

    this.validate(newValue, this.state.triedToSubmit)
      .then(errors => {
        this.setState({
          errors: containsError(errors) ? errors : new Immutable.Map()
        })
      })
  }

  onSubmit = (event) => {
    if (!this.props.disabled && !this.state.submitting) {
      this.validate(this.state.value, true)
        .then(errors => {
          // ...
        })
    }
  }
}
```

With that, `props.validate` would support a return value of either an object (like it is currently) or a `Promise` resolving with an object. Internally, `this.validate` would always be handled asynchronously and return a `Promise`.

I would like to talk about this in person and try it out together.

(Also, this would need to have new tests in "Form.spec.jsx".)

File: src/form/Form.jsx (hunk 1 above)
Comment (ingmarh) [reply src/form/Form.jsx]: Okay, now I see why you added this. If we end up extending `props.validate` to add async support, then I propose to switch the argument order to be `(value, forSubmit, changedField)` for backwards compatibility.

File: src/form/Form.jsx [hunk 3]
```diff
@@ -123,24 +147,34 @@ export default class Form extends React.Component {
...
-          this.setState({submitting: true})
-          result.catch(() => {}).then(() => this.setState({submitting: false}))
-        }
-      }
+  syncOrAsyncValidate = (value, changedFiled, triedToSubmit) => {
+    if (this.props.validateAsync) {
+      return this.linearizer(() => this.props.validateAsync(value, changedFiled, triedToSubmit))
+    } else {
+      return new Promise((resolve, reject) => {
+        resolve(this.props.validate(value, triedToSubmit))
```
Comment (Raigen) [inline src/form/Form.jsx]: You can simply return `Promise.resolve(...)`

**Review by Raigen — CHANGES_REQUESTED** [review]

**Review by ingmarh — APPROVED** [review]

**Review by Raigen — APPROVED** [review]

**Pull request merged**